pkg/metrics/metrics.go

```diff
@@ -248,6 +248,42 @@ func (c *Config) validate(ctx context.Context) error {
 	return nil
 }
 
+// RemoveStaleResources deletes stale resources that could be left behind during the
+// upgrade process due to the renaming of resources between current and previous operator version
+func (c *Config) RemoveStaleResources(ctx context.Context) {
+	operatorName, err := k8sutil.GetOperatorName()
+	if err != nil {
+		log.Error(err, "error getting operator name")
+	}
+
+	// staleResourceName is the metrics object name created for Prometheus monitoring by previous operator versions
+	staleResourceName := operatorName + "-metrics"
+
+	if service, err := c.CoreV1().Services(c.namespace).Get(ctx, staleResourceName, metav1.GetOptions{}); err == nil {
+		if err = c.CoreV1().Services(c.namespace).Delete(ctx, staleResourceName, metav1.DeleteOptions{}); err != nil {
+			c.Recorder.Eventf(service, core.EventTypeWarning, "serviceDeletionFailed",
+				"Service %s deletion failure", service.GetName())
+			log.Error(err, "error deleting stale service object")
+		}
+	}
+
+	if endpoint, err := c.CoreV1().Endpoints(c.namespace).Get(ctx, staleResourceName, metav1.GetOptions{}); err == nil {
+		if err = c.CoreV1().Endpoints(c.namespace).Delete(ctx, staleResourceName, metav1.DeleteOptions{}); err != nil {
+			c.Recorder.Eventf(endpoint, core.EventTypeWarning, "endpointDeletionFailed",
+				"Endpoint %s deletion failure", endpoint.GetName())
+			log.Error(err, "error deleting stale endpoint object")
+		}
+	}
+
+	if serviceMonitor, err := c.ServiceMonitors(c.namespace).Get(ctx, staleResourceName, metav1.GetOptions{}); err == nil {
+		if err = c.ServiceMonitors(c.namespace).Delete(ctx, staleResourceName, metav1.DeleteOptions{}); err != nil {
+			c.Recorder.Eventf(serviceMonitor, core.EventTypeWarning, "serviceMonitorDeletionFailed",
+				"ServiceMonitor %s deletion failure", serviceMonitor.GetName())
+			log.Error(err, "error deleting stale serviceMonitor object")
+		}
+	}
```

Comment:
What if the resource is present but the `Get` fails? Won't we be silently ignoring it? Furthermore, not attempting to delete the resource.

Reply:
I ll add a check for not found error.